accounts-db/src/io_uring/file_creator.rs

```diff
@@ -0,0 +1,550 @@
+
+        let op = FileCreatorOp::Open(OpenOp {
+            dir_handle,
+            path_bytes,
+            mode,
+            file_key,
+        });
+        self.ring.push(op)?;
+
+        Ok(file_key)
+    }
+
+    fn wait_add_file(&mut self, file: PendingFile) -> io::Result<usize> {
+        while self.ring.context().files.len() >= self.ring.context().files.capacity() {
+            self.ring.process_completions()?;
+            self.ring
+                .submit_and_wait(1, CHECK_PROGRESS_AFTER_SUBMIT_TIMEOUT)?;
+        }
+        let file_key = self.ring.context_mut().files.insert(file);
+        Ok(file_key)
+    }
+
+    fn write_and_close(&mut self, mut src: impl Read, file_key: usize) -> io::Result<()> {
+        let mut offset = 0;
+        loop {
+            let mut buf = self.wait_free_buf()?;
+
+            let state = self.ring.context_mut();
+            let file = &mut state.files[file_key];
+
+            let len = src.read(buf.as_mut())?;
+            if len == 0 {
+                file.eof = true;
+
+                state.buffers.push_front(buf);
+                if file.is_completed() {
+                    (state.file_complete)(mem::take(&mut file.path));
+                    self.ring
+                        .push(FileCreatorOp::Close(CloseOp::new(file_key)))?;
+                }
+                break;
+            }
+
+            file.writes_started += 1;
+            if file.completed_open {
+                let op = WriteOp {
+                    file_key,
+                    offset,
+                    buf,
+                    write_len: len,
+                };
+                state.submitted_writes_size += len;
+                self.ring.push(FileCreatorOp::Write(op))?;
+            } else {
+                file.backlog.push((buf, offset, len));
+            }
+
+            offset += len;
+        }
+
+        Ok(())
+    }
+
+    fn wait_free_buf(&mut self) -> io::Result<IoFixedBuffer> {
+        loop {
+            self.ring.process_completions()?;
+            let state = self.ring.context_mut();
+            if let Some(buf) = state.buffers.pop_front() {
+                return Ok(buf);
+            }
+            state.stats_no_buf_count += 1;
+            state.stats_no_buf_sum_submitted_write_sizes += state.submitted_writes_size;
+
+            self.ring
+                .submit_and_wait(1, CHECK_PROGRESS_AFTER_SUBMIT_TIMEOUT)?;
+        }
+    }
+}
+
+struct FileCreatorState<'a> {
+    files: Slab<PendingFile>,
+    buffers: VecDeque<IoFixedBuffer>,
+    /// Externally provided callback to be called on paths of files that were written
+    file_complete: Box<dyn FnMut(PathBuf) + 'a>,
+    open_fds: usize,
+    /// Total write length of submitted writes
+    submitted_writes_size: usize,
+    /// Count of cases when more than half of buffers are free (files are written
+    /// faster than submitted - consider less buffers or speeding up submission)
+    stats_large_buf_headroom_count: u32,
```

Comment:
nit: put these in a common structure instead of prefixing stats_ to all of them

Reply:
done